lib/Sema/DerivedConformances.cpp

```diff
@@ -340,3 +370,40 @@ DerivedConformance::declareDerivedProperty(TypeChecker &tc, Decl *parentDecl,
 
   return {propDecl, pbDecl};
 }
+
+bool DerivedConformance::checkAndDiagnoseDisallowedContext(
+    ValueDecl *synthesizing) const {
+  // In general, conformances can't be synthesized in extensions across files;
+  // but we have to allow it as a special case for Equatable and Hashable on
+  // enums with no associated values to preserve source compatibility.
+  bool allowCrossfileExtensions = false;
+  if (Protocol->isSpecificProtocol(KnownProtocolKind::Equatable) ||
+      Protocol->isSpecificProtocol(KnownProtocolKind::Hashable)) {
+    auto ED = dyn_cast<EnumDecl>(Nominal);
+    allowCrossfileExtensions = ED && ED->hasOnlyCasesWithoutAssociatedValues();
+  }
+
+  if (!allowCrossfileExtensions &&
+      Nominal->getModuleScopeContext() !=
+          getConformanceContext()->getModuleScopeContext()) {
+    TC.diagnose(ConformanceDecl->getLoc(),
+                diag::cannot_synthesize_in_crossfile_extension,
+                getProtocolType());
+    TC.diagnose(Nominal->getLoc(), diag::type_declared_here);
+    return true;
+  }
+
+  // A non-final class can't have an protocol-witnesss initializer in an
+  // extension.
+  if (auto CD = dyn_cast<ClassDecl>(Nominal)) {
+    if (!CD->isFinal() && isa<ConstructorDecl>(synthesizing) &&
+        isa<ExtensionDecl>(ConformanceDecl)) {
+      TC.diagnose(ConformanceDecl->getLoc(),
+                  diag::cannot_synthesize_init_in_extension_of_nonfinal,
+                  getProtocolType(), synthesizing->getFullName());
+      return true;
+    }
+  }
+
+  return false;
```

Comment:
In Swift 3 mode, we should disallow synthesis of conformances declared on extensions, because the Swift 3 meaning of `private` doesn't jive well with synthesis, e.g.,

```swift
struct X {
  private var foo: Int
}

extension X: Equatable { /* cannot access foo in Swift 3! */ }
```

Developers can migrate to Swift 4.x to get this feature.